File: tools/compare_go/generate_data.go
package main

import (
	"bufio"
	"encoding/json"
	"fmt"
	"log"
	"os"
	"sort"
	"strconv"
	"strings"
	"time"
)

//-----------------------------------------------------------------------------

// Generate the validation test data for AceTime using Go Lang.
//
// Usage:
//
// $ go run generate_data.go [--] [--help]
// [--start_year start]
// [--until_year until]
// [--epoch_year year]
// < zones.txt
// > validation_data.json
func main() {
	parseArgs()
	dt := time.Date(epochYear, 1, 1, 0, 0, 0, 0, time.UTC)
	secondsToAceTimeEpochFromUnixEpoch = dt.Unix()

	zones := readZones()
	testData := processZones(zones)
	validationData := createValidationData(testData)
	printJson(validationData)
}

//-----------------------------------------------------------------------------

// Default values of various flags.
var (
	startYear        = 2000
	untilYear        = 2100
	epochYear        = 2050
	samplingInterval = 22 // hours

	// Number of seconds from Unix Epoch (1970-01-01 00:00:00) to AceTime Epoch
	// (2050-01-01 00:00:00), configurable using --epoch_year.
	secondsToAceTimeEpochFromUnixEpoch int64
)

func usage() {
	fmt.Println(`Usage: go run generate_data.go [--] [--help]
        [--sampling_interval hours] [--start_year start] [--until_year until]
				[--epoch_year year]
        < zones.txt > validation_data.json`)
}

// ParseArgs() is a function that parses the command line arguments in os.Args,
// looking for optional long flags (beginning with two dashes), optional short
// flags (beginning with one dash), and positional arguments at the end. Returns
// a slice of the remaining positional arguments which maybe empty slice.
//
// There is probably std library module that already handles command line
// argument parsing. I rolled my own to help me learn the Go lang.
func parseArgs() []string {
	// Shift out the name of the command in Args[0]
	args := os.Args[1:]

	var err error

	// Loop through each command line argument, looking for optional flags,
	// and terminating if a positional argument is found.
	for len(args) > 0 {
		s := args[0]
		if s == "--start_year" {
			args = args[1:]
			if len(args) == 0 {
				fmt.Println("--start_year must have an argument")
				os.Exit(1)
			}
			s = args[0]
			startYear, err = strconv.Atoi(s)
			if err != nil {
				fmt.Println("--start_year must have an integer value")
				os.Exit(1)
			}
		} else if s == "--until_year" {
			args = args[1:]
			if len(args) == 0 {
				fmt.Println("--until_year must have an argument")
				os.Exit(1)
			}
			s = args[0]
			untilYear, err = strconv.Atoi(s)
			if err != nil {
				fmt.Println("--until_year must have an integer value")
				os.Exit(1)
			}
		} else if s == "--epoch_year" {
			args = args[1:]
			if len(args) == 0 {
				fmt.Println("--epoch_year must have an argument")
				os.Exit(1)
			}
			s = args[0]
			epochYear, err = strconv.Atoi(s)
			if err != nil {
				fmt.Println("--epoch_year must have an integer value")
				os.Exit(1)
			}
		} else if s == "--sampling_interval" {
			args = args[1:]
			if len(args) == 0 {
				fmt.Println("--sampling_interval must have an argument")
				os.Exit(1)
			}
			s = args[0]
			samplingInterval, err = strconv.Atoi(s)
			if err != nil {
				fmt.Println("--sampling_interval must have an integer value")
				os.Exit(1)
			}
		} else if s == "--help" {
			usage()
			os.Exit(0)
		} else if s == "--" {
			args = args[1:]
			break
		} else if len(s) > 0 && s[0] == '-' {
			fmt.Printf("Invalid flag '%s'\n", s)
			os.Exit(1)
		} else {
			break
		}

		args = args[1:]
	}

	return args
}

//-----------------------------------------------------------------------------

// The collection of all zones and their test items.
type ValidationDataType struct {
	StartYear      int          `json:"start_year"`
	UntilYear      int          `json:"until_year"`
	EpochYear      int          `json:"epoch_year"`
	Source         string       `json:"source"`
	Version        string       `json:"version"`
	TzVersion      string       `json:"tz_version"`
	HasValidAbbrev bool         `json:"has_valid_abbrev"`
	HasValidDst    bool         `json:"has_valid_dst"`
	TestData       TestDataType `json:"test_data"`
}

// A map of zoneName to an array of TestItems.
type TestDataType map[string][]TestItemType

// A test item struct contains the epochSeconds with its local date time
// components.
type TestItemType struct {
	EpochSeconds int    `json:"epoch"`
	TotalOffset  int    `json:"total_offset"`
	DstOffset    int    `json:"dst_offset"`
	Year         int    `json:"y"`
	Month        int    `json:"M"`
	Day          int    `json:"d"`
	Hour         int    `json:"h"`
	Minute       int    `json:"m"`
	Second       int    `json:"s"`
	Abbrev       string `json:"abbrev"`
	ItemType     string `json:"type"` // "A", "B", "S", "T" or "Y"
}

//-----------------------------------------------------------------------------

// ReadZones reads the list of zones from the stdin and returns them in an array
// of strings.
func readZones() []string {
	zones := make([]string, 0, 600)

	reader := bufio.NewReader(os.Stdin)
	for {
		message, err := reader.ReadString('\n')
		if err != nil {
			break
		}

		// Skip comments or blank lines.
		trimmed := strings.Trim(message, " \t\n")
		size := len(trimmed)
		if size == 0 || trimmed[0] == '#' {
			continue
		}

		// Extract zone name.
		message = strings.TrimRight(message, "\n")
		zones = append(zones, message)
	}

	return zones
}

//-----------------------------------------------------------------------------

// ProcessZones() iterates over the list of zones and calls processZone().
func processZones(zones []string) TestDataType {
	testData := make(TestDataType)
	for _, zoneName := range zones {
		testItems, err := processZone(zoneName)
		if err != nil {
			fmt.Printf("Unable to process zone '%s'\n", zoneName)
			continue
		}

		testData[zoneName] = testItems
	}

	return testData
}

// ProcessZone() adds the DST transitions and sample data for each month.
func processZone(zoneName string) ([]TestItemType, error) {
	tz, err := time.LoadLocation(zoneName)
	if err != nil {
		return nil, err
	}

	testItems := make([]TestItemType, 0, 500)
	testItems = addTransitions(testItems, tz)
	testItems = addSamples(testItems, tz)
	sortSamples(testItems)
	return testItems, nil
}

func sortSamples(testItems []TestItemType) {
	sort.Slice(testItems, func(i, j int) bool {
		return testItems[i].EpochSeconds < testItems[j].EpochSeconds
	})
}

// AddTransition() finds the DST transitions of the timezone, and creates
// a testItem for the 'before' time, and a testItem for the 'after' time.
func addTransitions(
	testItems []TestItemType,
	tz *time.Location,
) []TestItemType {

	transitions := findTransitions(startYear, untilYear, samplingInterval, tz)
	for _, transition := range transitions {
		beforeTestItem := createTestItem(transition.before, "A")
		afterTestItem := createTestItem(transition.after, "B")
		testItems = append(testItems, beforeTestItem)
		testItems = append(testItems, afterTestItem)
	}

	return testItems
}

// AddSamples() add a testItem for each month for the given zone.
func addSamples(
	testItems []TestItemType,
	tz *time.Location,
) []TestItemType {
	for year := startYear; year < untilYear; year++ {
		for month := 1; month <= 12; month++ {
			dtSample := time.Date(year, time.Month(month), 2, 0, 0, 0, 0, tz)
			sampleTestItem := createTestItem(dtSample, "S")
			testItems = append(testItems, sampleTestItem)
		}
	}

	return testItems
}

func createTestItem(t time.Time, itemType string) TestItemType {
	unixSeconds := t.Unix()
	epochSeconds := int(unixSeconds - secondsToAceTimeEpochFromUnixEpoch)

	// seconds
	totalOffset := utcOffsetMinutes(t) * 60
	// Go lang time package does not provide the DST offset.
	dstOffset := 0
	abbrev := abbreviation(t)

	return TestItemType{
		epochSeconds,
		totalOffset,
		dstOffset,
		t.Year(),
		int(t.Month()),
		t.Day(),
		t.Hour(),
		t.Minute(),
		t.Second(),
		abbrev,
		itemType,
	}
}

func abbreviation(t time.Time) string {
	abbrev := t.Format("MST")
	return abbrev
}

//-----------------------------------------------------------------------------

func createValidationData(testData TestDataType) ValidationDataType {
	return ValidationDataType{
		StartYear:      startYear,
		UntilYear:      untilYear,
		EpochYear:      epochYear,
		Source:         "go",
		Version:        "",
		TzVersion:      "",
		HasValidAbbrev: true,
		HasValidDst:    false,
		TestData:       testData,
	}
}

//-----------------------------------------------------------------------------

func printJson(validationData ValidationDataType) {
	jsonData, err := json.MarshalIndent(validationData, "", "  ")
	if err != nil {
		log.Println(err)
	}
	fmt.Println(string(jsonData))
}

File: tools/compare_acetimec/generate_data.c
/*
 * Generate the validation JSON output for the zones given on the STDIN. The
 * transition time and UTC offsets are calculated using the AceTimeC library
 * (https://github.com/bxparks/AceTimeC).
 *
 * Usage:
 * $ ./generate_data.out
 *    [--start_year start]
 *    [--until_year until]
 *    < zones.txt
 *    > validation_data.json
 */

#include <stdbool.h>
#include <stdlib.h> // exit(), qsort()
#include <string.h> // strcmp(), strncmp()
#include <stdio.h> // printf(), fprintf()
#include <acetimec.h>
#include "test_data.h"
#include "sampling.h"

// Command line arguments
int16_t start_year = 2000;
int16_t until_year = 2050;

bool is_registry_sorted;

//-----------------------------------------------------------------------------

static int compare_test_item(const void *a, const void *b)
{
  const struct TestItem *ta = a;
  const struct TestItem *tb = b;
  if (ta->epoch_seconds < tb->epoch_seconds) return -1;
  if (ta->epoch_seconds > tb->epoch_seconds) return 1;
  return 0;
}

/** Sort the TestItems of each TestDataEntry according to epochSeconds. */
void sort_test_data(struct TestData *test_data)
{
  for (int i = 0; i < test_data->num_entries; i++) {
    struct TestDataEntry *entry = &test_data->entries[i];
    qsort(
        entry->items,
        entry->num_items,
        sizeof(struct TestItem),
        compare_test_item);
  }
}

/**
 * Generate the JSON output on STDOUT which will be redirect into
 * 'validation_data.json' file. Adopted from GenerateData.java.
 */
void print_json(const struct TestData *test_data) {
  const char indent0[] = "  ";
  const char indent1[] = "    ";
  const char indent2[] = "      ";
  const char indent3[] = "        ";

  printf("{\n");
  printf("%s\"start_year\": %d,\n", indent0, start_year);
  printf("%s\"until_year\": %d,\n", indent0, until_year);
  printf("%s\"source\": \"AceTimeC\",\n", indent0);
  printf("%s\"version\": \"%s\",\n", indent0, ACE_TIME_C_VERSION_STRING);
  printf("%s\"tz_version\": \"%s\",\n", indent0, kAtcTzDatabaseVersion);
  printf("%s\"has_valid_abbrev\": true,\n", indent0);
  printf("%s\"has_valid_dst\": true,\n", indent0);
  printf("%s\"test_data\": {\n", indent0);

  // Print each zone
  int num_zones = test_data->num_entries;
  for (int z = 0; z < num_zones; z++) {
    const struct TestDataEntry *entry = &test_data->entries[z];
    const char *zone_name = entry->zone_name;
    printf("%s\"%s\": [\n", indent1, zone_name);

    // Print each testItem
    for (int i = 0; i < entry->num_items; i++) {
      const struct TestItem *item = &entry->items[i];
      printf("%s{\n", indent2);
      {
        printf("%s\"epoch\": %ld,\n", indent3, item->epoch_seconds);
        printf("%s\"total_offset\": %d,\n", indent3, item->utc_offset);
        printf("%s\"dst_offset\": %d,\n", indent3, item->dst_offset);
        printf("%s\"y\": %d,\n", indent3, item->year);
        printf("%s\"M\": %d,\n", indent3, item->month);
        printf("%s\"d\": %d,\n", indent3, item->day);
        printf("%s\"h\": %d,\n", indent3, item->hour);
        printf("%s\"m\": %d,\n", indent3, item->minute);
        printf("%s\"s\": %d,\n", indent3, item->second);
        printf("%s\"abbrev\": \"%s\",\n", indent3, item->abbrev);
        printf("%s\"type\": \"%c\"\n", indent3, item->type);
      }
      printf("%s}%s\n", indent2, (i < entry->num_items - 1) ? "," : "");
    }

    printf("%s]%s\n", indent1, (z < test_data->num_entries - 1) ? "," : "");
  }

  printf("%s}\n", indent0);
  printf("}\n");
}

/** Insert TestItems for the given 'zone_name' into test_data. */
bool process_zone(
    struct AtcZoneProcessing *processing,
    struct TestData *test_data,
    const char *zone_name)
{
  const struct AtcZoneInfo *zone_info = atc_registrar_find_by_name(
      kAtcZoneAndLinkRegistry,
      kAtcZoneAndLinkRegistrySize,
      zone_name,
      is_registry_sorted);

  if (zone_info == NULL) {
    fprintf(stderr, "ERROR: Zone %s: not found\n", zone_name);
    return false;
  }

  struct TestDataEntry *entry = test_data_next_entry(test_data);
  add_transitions(
      entry,
      zone_name,
      processing,
      zone_info,
      start_year,
      until_year);
  add_monthly_samples(
      entry,
      zone_name,
      processing,
      zone_info,
      start_year,
      until_year);
  return true;
}

/**
 * Read the list of zones from the 'zones.txt' in the stdin. Ignore blank lines
 * and comments (starting with '#'), and process each zone, one per line.
 */
bool read_and_process_zone(
    struct AtcZoneProcessing *processing,
    struct TestData *test_data)
{
  char line[MAX_LINE_LENGTH];
  while (1) {
    char *p = fgets(line, MAX_LINE_LENGTH, stdin);
    if (! p) break;
    const char *delim = " ";
    char *word = strtok(line, delim);
    // Strip trailing \n
    size_t len = strlen(word);
    if (len && word[len-1] == '\n') {
      word[len-1] = '\0';
    }
    if (strcmp(word, "") == 0) continue;
    if (word[0] == '#') continue;

    bool status = process_zone(processing, test_data, word);
    if (! status) return status;
  }

  return true;
}

//-----------------------------------------------------------------------------

void usage_and_exit() {
  fprintf(stderr,
    "Usage: generate_data [--install_dir {dir}]\n"
    "   [--start_year start] [--until_year until]\n"
    "   < zones.txt\n");
  exit(1);
}

#define SHIFT(argc, argv) \
  do { \
    argc--; \
    argv++; \
  } while (0)

static bool argEquals(const char *s, const char *t) {
  return strcmp(s, t) == 0;
}

int main(int argc, const char* const* argv) {
  // Parse command line flags.
  const char *start = "2000";
  const char *until = "2050";

  SHIFT(argc, argv);
  while (argc > 0) {
    if (argEquals(argv[0], "--start_year")) {
      SHIFT(argc, argv);
      if (argc == 0) usage_and_exit();
      start = argv[0];
    } else if (argEquals(argv[0], "--until_year")) {
      SHIFT(argc, argv);
      if (argc == 0) usage_and_exit();
      until = argv[0];
    } else if (argEquals(argv[0], "--")) {
      SHIFT(argc, argv);
      break;
    } else if (strncmp(argv[0], "-", 1) == 0) {
      fprintf(stderr, "Unknonwn flag '%s'\n", argv[0]);
      usage_and_exit();
    } else {
      break;
    }
    SHIFT(argc, argv);
  }

  start_year = atoi(start);
  until_year = atoi(until);

  // Set up registry
  is_registry_sorted = atc_registrar_is_registry_sorted(
      kAtcZoneAndLinkRegistry,
      kAtcZoneAndLinkRegistrySize);

  // Initialize an AtcZoneProcessing instance.
  struct AtcZoneProcessing processing;
  atc_processing_init(&processing);

  // Process the zones on the STDIN
  struct TestData test_data;
  test_data_init(&test_data);
  bool status = read_and_process_zone(&processing, &test_data);
  if (! status) exit(1);

  sort_test_data(&test_data);
  print_json(&test_data);
  test_data_free(&test_data);

  return 0;
}
